enigma-principal/app/src/controller/km_http_server.rs

```diff
@@ -1,184 +1,167 @@
-        let addrs = epoch_state.get_contract_addresses(&worker.into())?;
-        Ok(addrs)
+        Ok(epoch_state.get_contract_addresses(&worker.into())?)
     }
 
     #[logfn(DEBUG)]
-    pub fn get_state_keys(epoch_provider: &EpochProvider, request: StateKeyRequest) -> Result<Value, Error> {
-        let epoch_state = match request.block_number.clone() {
-            Some(block_number) => epoch_provider.find_epoch(block_number.try_into()?)?,
-            None => epoch_provider.find_last_epoch()?,
+    pub fn get_state_keys(&self, request: StateKeyRequest) -> Result<Value, ControllerError> {
+        let signed_epoch = match request.block_number {
+            Some(block_number) => self.controller.find_epoch(block_number)?,
+            None => self.controller.find_last_epoch()?,
         };
-        let addresses = &request.addresses;
-        let addrs: Vec<ContractAddress> = {
-            if let Some(addrs) = addresses {
-                let res: Result<Vec<ContractAddress>, _>  = addrs.iter().map(|item| ContractAddress::from_hex(item)).collect();
-                res?
-            }
-            else{
-                let msg = PrincipalMessage::from_message(&request.get_data()?)?;
-                Self::find_epoch_contract_addresses(&request, &msg, &epoch_state)?
-            }
+        let addrs = match request.addresses.clone() {
+            Some(addrs) => addrs,
+            None => {
+                let msg = PrincipalMessage::from_message(&request.get_data()).
+                    map_err(|_| ControllerError::HTTPServerError(HTTPServerError::MessagingErr))?;
+                Self::find_epoch_contract_addresses(&request, &msg, &signed_epoch)?
+            },
         };
-        let response = get_enc_state_keys(*epoch_provider.eid, request, epoch_state.nonce, &addrs)?;
-        let response_data = serde_json::to_value(&response)?;
+        let response =
+            get_enc_state_keys(self.controller.eid, request, signed_epoch.get_nonce(), &addrs).
+            map_err(ControllerError::EnclaveError)?;
+        let response_data = serde_json::to_value(&response).
+            map_err(|_| ControllerError::HTTPServerError(HTTPServerError::MessagingErr))?;
         Ok(response_data)
     }
 
-    fn handle_error(internal_err: Error) -> ServerError {
-        if let Some(err) = internal_err.downcast_ref::<EnclaveFailError>() {
-            error!("{:?}", internal_err.as_fail());
-            let server_err = match &err.err {
-                EnclaveReturn::WorkerAuthError => {
-                    ServerError {
-                        code: ErrorCode::ServerError(JSON_RPC_ERROR_WORKER_NOT_AUTHORIZED),
-                        message: format!("Worker not authorized to request the keys: {:?}.", err),
-                        data: None,
-                    }
+    fn handle_error(internal_err: ControllerError) -> ServerError {
+        match internal_err {
+            ControllerError::EnclaveError(e) => {
```

Comment:
This giant match statement can be replaced by:
```rust
fn handle_error(internal_err: ControllerError) -> ServerError {
    match internal_err {
        ControllerError::EnclaveError(
            EnclaveError::EnclaveFailErr {
                err: e @ EnclaveReturn::WorkerAuthError,
                ..
            }
        ) => ServerError {
            code: ErrorCode::ServerError(JSON_RPC_ERROR_WORKER_NOT_AUTHORIZED),
            message: format!("Worker not authorized to request the keys: {:?}.", e),
            data: None,
        },
        ControllerError::EnclaveError(
            EnclaveError::EnclaveFailErr { err: e @ _, .. }
        ) => ServerError {
            code: ErrorCode::InternalError,
            message: format!("Internal error in enclave: {:?}", e),
            data: None,
        },
        ControllerError::EnclaveError(e @ _) => ServerError {
            code: ErrorCode::InternalError,
            message: format!("Internal error: {:?}", e),
            data: None,
        },
        _ => ServerError {
            code: ErrorCode::InternalError,
            message: format!("Internal error: {:?}", internal_err),
            data: None,
        }
    }
}
```

And even further reduced by having each arm return a tuple of `(code, message)` and then build the `ServerError` once after the match